authprox/service.go

```diff
@@ -0,0 +1,354 @@
+				Public: out.LongPubs[0],
+			})
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &resp, nil
+}
+
+// Enroll will save the proposed secret key share (and associated information)
+// in the local database.
+func (s *service) Enroll(req *EnrollRequest) (*EnrollResponse, error) {
+	// TODO: How is auth for the right to enroll done?
+	// idea: allow other services (i.e. evoting) to define certain type/claims
+	// as admins, require AuthInfo in EnrollmentRequest, and if the validated
+	// claim is in the admin list, allow.
+
+	// Prepare the stuff to be written
+	k, err := protobuf.Encode(&ti{T: req.Type, I: req.Issuer})
+	if err != nil {
+		return nil, err
+	}
+	lpri := share.PriShare{
+		I: req.LongPri.I,
+		V: req.LongPri.V,
+	}
+	v, err := protobuf.Encode(&dssConfig{
+		Participants: req.Participants,
+		LongPri:      lpri,
+		LongPubs:     req.LongPubs,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Write the type/claim -> dssConfg into the database.
+	err = s.db.Update(func(tx *bolt.Tx) error {
+		// Need to do the find inside of the Update tx, or else it is racy
+		// with respect to other writers.
+		b := tx.Bucket(s.bucket)
+		if b == nil {
+			return errors.New("nil bucket")
+		}
+		if ret := b.Get(k); ret != nil {
+			return fmt.Errorf("enrollment already exists for type:issuer %v:%v", req.Type, req.Issuer)
+		}
+		return b.Put(k, v)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &EnrollResponse{}, nil
+}
+
+// Signature will verify the authentication information in
+// the request, according to the rules specific to that authentication type.
+// If the information is valid, it will then generate a partial signature
+// on a new message which binds together the claim found from the authentication
+// information, and the message in the request, using the secret key share associated
+// with the type and issuer. It is the caller's responsibility to gather a
+// threshold of key shares and then combine them into a final signature.
+func (s *service) Signature(req *SignatureRequest) (*SignatureResponse, error) {
+	if req == nil {
+		return nil, errors.New("no request")
+	}
+
+	// Look for a validator for the external type requested.
+	validator, err := s.validator(req.Type)
+	if err != nil {
+		return nil, err
+	}
+
+	// Use the validator to extract a claim from the auth info.
+	claim, _, err := validator.FindClaim(req.Issuer, req.AuthInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: check that if extraData != "", then extraData == msg
+
+	// Given this claim, sign a
+	h := sha256.New()
+	b := make([]byte, 4)
+	binary.LittleEndian.PutUint32(b, uint32(len(claim)))
+	h.Write(b)
+	h.Write([]byte(claim))
+	h.Write(req.Message)
+	msg2 := h.Sum(nil)
+
+	// Find the config that is associated with the issuer.
+	dsscfg, err := s.find(req.Type, req.Issuer)
+	if err != nil {
+		return nil, fmt.Errorf("cannot find key: %v", err)
+	}
+
+	// Make the signature.
+	suite := suites.MustFind("ed25519")
```

Comment:
Why don't you use `cothority.Suite` here?

Reply:
Because kyber/sign/dss makes EdDSA compatible signatures, and thus only works with ed25519. That's the same reason the rule format is `proxy:abcde:user@example.com` and not `proxy:ed25519:abcde:user@example.com`.